# Exploring eBay Car Sales Data

In this project, I'll work with a dataset of used cars from eBay Kleinanzeigen, a classfieds section of German eBay website.

The dataset was originally scraped and uploaded to Kaggle by user [orgesleka](https://www.kaggle.com/orgesleka). The original dataset isn't available on Kaggle anymore, but it can be found [here](https://data.world/data-society/used-cars-data).

The goal for this project is to clean the dataset and analyze the included used car listings, it is to be done to familiarize myself with using Python pandas.

### Importing libraries and read the data in dataframe

In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv("autos.csv", encoding = "Latin-1")

In [2]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

The dataset contains 20 columns, most of which are strings.

Some column titles are [CamelCase](https://en.wikipedia.org/wiki/Camel_case), instead of Python's preferred [snakecase](https://en.wikipedia.org/wiki/Snake_case), some changes need to be made to suit the purpose of data cleaning.

It is noticeable after redenring the `autos.info()`, under the `Non-Null Count` column, quite some rows contain `NaN` values and they are all appeared to be `object`,  which we will need to assess them later on.   

### Cleaning column names

In [4]:
autos.rename(columns = {"yearOfRegistration": "registration_year", "monthOfRegistration":"registration_month",
                       "notRepairedDamage":"unrepaired_damage","dateCreated":"ad_created", "dateCrawled":"date_crawled",
                       "offerType":"offer_type","vehicleType":"vehicle_type","fuelType":"fuel_type","postalCode":"postal_code",
                       "lastSeen":"last_seen"}, inplace = True)

In [5]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  powerPS   model   odometer  registration_month fuel_type  \
0    manuell      158  andere  150,000km                   3       lpg   
1  automatik      286     7er  150,000km                   6    benzin   
2    manuell      102    golf   70,000km                   7    benzin   
3  automatik       71  fortwo   70,000km                   6    benzin   
4    manuell        0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Previously, the `yearOfRegistration`, `monthOfRegistration`, `notRepairedDamage`, and `dateCreated` columns are camelCased, and there is no space between the words which might essentially lead to confusion when working with large dataset. Thus, they are changed to `registration_year`, `registration_month`, `unrepaired_damage`, and `ad_created`, respectively. 

### Initial Exploration and Cleaning

Next step is to work on some basic data explration to determine what other cleaning tasks need to be performed, we can look for:
* Text columns where all or almost all values are the same, this means the column does not contain useful information for analysis.
* Example of numeric data being stored as text which can be cleaned and converted.

In [6]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox       powerPS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nrOfPictures   postal_code            last_seen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

After reviewing the `autos` dataset one more time, all columns consist at least 2 or more unique values except `nrOfPictures` in which has `NaN` for every row value due to `unique` value equals to `NaN`, the rest looks good for now, we will assess them more closely before deciding to drop any unuseful columns. 

Notice `odometer` column, it is reported as string make up by some number in front and 'km' at the end, we can perform some cleaning on this column. 

Same goes for `price` column, since it is made up of a `$` followerd with some number, we will need to separate the `$` and turn the values into `int` or `float`.

#### Price Column Cleaning

In [7]:
autos['price'] = autos['price'].str.replace('[$,]','', regex = True)
#autos['price'] = autos['price'].str.replace(',','', regex = True)

In [8]:
autos['price'] = autos['price'].astype(int)
autos['price'].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int32

#### Odometer Column Cleaning

In [9]:
autos['odometer'] = autos['odometer'].str.replace('[a-z,]*','', regex = True)
#autos['price'] = autos['price'].str.replace(',','', regex = True)

In [10]:
autos['odometer'] = autos['odometer'].astype(int)
autos['odometer'].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer, dtype: int32

#### Column Name Changed

In [11]:
autos.rename(columns = {'price':'price_dollar',"odometer":"odometer_km"},inplace = True)

In [12]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price_dollar   abtest vehicle_type  registration_year  \
0  privat    Angebot          5000  control          bus               2004   
1  privat    Angebot          8500  control    limousine               1997   
2  privat    Angebot          8990     test    limousine               2009   
3  privat    Angebot          4350  control   kleinwagen               2007   
4  privat    Angebot          1350     test        kombi               2003   

     gearbox  powerPS   model  odometer_km  registration_month fuel_type  \
0    manuell      158  andere       150000                   3       lpg   
1  automatik      286     7er       150000                   6    benzin   
2    manuell      102    golf        70000                   7    benzin   
3  automatik       71  fortwo        70000                   6    benzin   
4    manuell        0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

#### Exploring the Odometer and Price Columns

In this step, we are going to look for data that is not necessary fit in the dataset (outliers).

##### Price

In [13]:
autos['price_dollar'].unique().shape

(2357,)

In [14]:
autos['price_dollar'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_dollar, dtype: float64

In [15]:
autos['price_dollar'].value_counts().sort_index(ascending = False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price_dollar, dtype: int64

We can notice that there are some absurd price such as '123456', '11111111', or '99999999', given the context of a market for used cars. It is acceptable to believe that these price are totally wrong and that we need to remove them from the dataset. We will mark the appropriate price range between '0' and '350000', although some might be false, we are just giving the benefit of the doubt here because we cannot rely on the `standard deviation` of '481104', which greatly exceeds the mean, '9840'. 

In [16]:
x = 0
y = 350000
autos = autos[ autos['price_dollar'].between(0, 350000)]
autos['price_dollar'].describe()

count     49986.000000
mean       5721.525167
std        8983.617820
min           0.000000
25%        1100.000000
50%        2950.000000
75%        7200.000000
max      350000.000000
Name: price_dollar, dtype: float64

##### Odometer 

In [17]:
autos['odometer_km'].unique().shape

(13,)

In [18]:
autos['odometer_km'].describe()

count     49986.000000
mean     125736.506222
std       40038.133399
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [19]:
autos['odometer_km'].value_counts().sort_index(ascending = False)

150000    32416
125000     5169
100000     2168
90000      1757
80000      1436
70000      1230
60000      1164
50000      1025
40000       818
30000       789
20000       784
10000       264
5000        966
Name: odometer_km, dtype: int64

It is apparent that `odometer_km` column does not contain outliers. Techniquely, an outlier is idetified if it exceeds 3 or more times the `standard deviation` from the `mean`, in the `odometer_km` case, if the value falls outside (5736 , 245736), then it would be outliers. However, we can see quite lots of values reported at '5000' (966), thus, it'd be better to keep them in the dataset otherwise we might risk losing information.

### Exploring the Date Column

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differiate by referring to the data dictionary.
- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

Right now, the `date_crawled`, `last_seen`, and `ad_created` columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively. The other two columns are represented as numeric values, so we can use methods like `Series.describe()` to understand the distribution without any extra data processing.

In [20]:
autos['date_crawled'].str[:10].value_counts(normalize = True, dropna = False).sort_index()

2016-03-05    0.025387
2016-03-06    0.013944
2016-03-07    0.035970
2016-03-08    0.033269
2016-03-09    0.033209
2016-03-10    0.032129
2016-03-11    0.032489
2016-03-12    0.036770
2016-03-13    0.015564
2016-03-14    0.036630
2016-03-15    0.033990
2016-03-16    0.029508
2016-03-17    0.031509
2016-03-18    0.013064
2016-03-19    0.034910
2016-03-20    0.037831
2016-03-21    0.037490
2016-03-22    0.032909
2016-03-23    0.032389
2016-03-24    0.029108
2016-03-25    0.031749
2016-03-26    0.032489
2016-03-27    0.031049
2016-03-28    0.034850
2016-03-29    0.034150
2016-03-30    0.033629
2016-03-31    0.031909
2016-04-01    0.033809
2016-04-02    0.035410
2016-04-03    0.038691
2016-04-04    0.036490
2016-04-05    0.013104
2016-04-06    0.003181
2016-04-07    0.001420
Name: date_crawled, dtype: float64

In [21]:
autos['date_crawled'].str[:10].value_counts(normalize = True, dropna = False).sort_values()

2016-04-07    0.001420
2016-04-06    0.003181
2016-03-18    0.013064
2016-04-05    0.013104
2016-03-06    0.013944
2016-03-13    0.015564
2016-03-05    0.025387
2016-03-24    0.029108
2016-03-16    0.029508
2016-03-27    0.031049
2016-03-17    0.031509
2016-03-25    0.031749
2016-03-31    0.031909
2016-03-10    0.032129
2016-03-23    0.032389
2016-03-11    0.032489
2016-03-26    0.032489
2016-03-22    0.032909
2016-03-09    0.033209
2016-03-08    0.033269
2016-03-30    0.033629
2016-04-01    0.033809
2016-03-15    0.033990
2016-03-29    0.034150
2016-03-28    0.034850
2016-03-19    0.034910
2016-04-02    0.035410
2016-03-07    0.035970
2016-04-04    0.036490
2016-03-14    0.036630
2016-03-12    0.036770
2016-03-21    0.037490
2016-03-20    0.037831
2016-04-03    0.038691
Name: date_crawled, dtype: float64

The site was crawled daily between for a month from March, 2016 to April, 2016.  Additionally, the amount spent each day crawling data is roughly normally distributed.

In [22]:
(autos['ad_created'].str[:10].value_counts(normalize = True, dropna = False).sort_index())

2015-06-11    0.000020
2015-08-10    0.000020
2015-09-09    0.000020
2015-11-10    0.000020
2015-12-05    0.000020
                ...   
2016-04-03    0.038931
2016-04-04    0.036850
2016-04-05    0.011843
2016-04-06    0.003261
2016-04-07    0.001280
Name: ad_created, Length: 76, dtype: float64

The `ad_created` columns contains more variety of dates that the car was listed, ranging from 1-2 months up to 10 months back.

In [23]:
(autos['last_seen'].str[:10]
                .value_counts(normalize = True, dropna = False)
                .sort_index()
)

2016-03-05    0.001080
2016-03-06    0.004421
2016-03-07    0.005362
2016-03-08    0.007582
2016-03-09    0.009843
2016-03-10    0.010763
2016-03-11    0.012524
2016-03-12    0.023807
2016-03-13    0.008983
2016-03-14    0.012804
2016-03-15    0.015884
2016-03-16    0.016445
2016-03-17    0.027928
2016-03-18    0.007422
2016-03-19    0.015744
2016-03-20    0.020706
2016-03-21    0.020726
2016-03-22    0.021586
2016-03-23    0.018585
2016-03-24    0.019565
2016-03-25    0.019205
2016-03-26    0.016965
2016-03-27    0.016024
2016-03-28    0.020846
2016-03-29    0.022326
2016-03-30    0.024847
2016-03-31    0.023827
2016-04-01    0.023106
2016-04-02    0.024887
2016-04-03    0.025367
2016-04-04    0.024627
2016-04-05    0.124275
2016-04-06    0.220982
2016-04-07    0.130957
Name: last_seen, dtype: float64

This is the date the crawler last saw any listing. It indicates to us the the date that listing was removed from the site, it is safe to assume that the car is being sold.

The last three days of the month contain disproportionate amount of 'last_seen' values. Due to the unsual 10 times the rest of the days, it is very unlikely that car sales spiked through the roof, and more likely that it was the ending of crawling period.

In [24]:
(autos['registration_year'].describe())

count    49986.000000
mean      2005.075721
std        105.727161
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Right of the bat, the min and max values of the `registration_year` column on the year the car was registered, catch our eyes. `1000` and `9999` is simply wrong and we will need to omit these values later.

#### Dealing with Incorrect Registration Year

Because car cannot be first registered before the latest listing was seen, therefore, any registration year after 2016 is inaccurate. Determining the earliest registered year is tricky, however, realistically we will just make it to be `1900`. 

One way to understand how to deal with inaccurate data is to know their portion in the dataset.

In [25]:
(~autos['registration_year'].between(1900,2016)).sum() / autos.shape[0]

0.03941103508982515

Because the registered year of car earlier than `1900` and later than `2016` is roughly 4%, we can omit the rows containing these years. 

In [26]:
autos = autos[autos['registration_year'].between(1900,2016)]
autos["registration_year"].value_counts().sort_index().head(15)

1910    9
1927    1
1929    1
1931    1
1934    2
1937    4
1938    1
1939    1
1941    2
1943    1
1948    1
1950    3
1951    2
1952    1
1953    1
Name: registration_year, dtype: int64

It seems that there are quite some cars registered in the early 90s.

### Exploring Price by Brand

In [27]:
autos["brand"].value_counts(normalize=True)

volkswagen        0.212117
bmw               0.110026
opel              0.108172
mercedes_benz     0.095364
audi              0.086409
ford              0.069768
renault           0.047359
peugeot           0.029532
fiat              0.025866
seat              0.018181
skoda             0.016036
mazda             0.015141
nissan            0.015099
smart             0.013912
citroen           0.013912
toyota            0.012475
sonstige_autos    0.010892
hyundai           0.009851
volvo             0.009247
mini              0.008643
mitsubishi        0.008143
honda             0.007852
kia               0.007102
alfa_romeo        0.006623
porsche           0.006102
suzuki            0.005915
chevrolet         0.005706
chrysler          0.003665
daihatsu          0.002562
dacia             0.002562
jeep              0.002249
subaru            0.002187
land_rover        0.002041
saab              0.001604
jaguar            0.001583
trabant           0.001562
daewoo            0.001500
r

German manufactorers appear 4 times in top 5 most car listings on the site. Volkswagen is by far the most popular brand that appears on the site, approximately double the car sales of the next two brands combined.

There are quite some brands that do not have significant amount of listings. Therefore, we are only going to include brands that has greater than 5% of listings.

In [28]:
brand_counts = autos['brand'].value_counts(normalize = True)
common_brand = brand_counts[ brand_counts > 0.05].index
print(common_brand)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [29]:
brand_mean_prices = {}

for brand in common_brand:
    mean_price = autos[autos['brand'] == brand]['price_dollar'].mean()
    brand_mean_prices[brand] = int(mean_price)
    
brand_mean_prices

{'volkswagen': 5231,
 'bmw': 8102,
 'opel': 2876,
 'mercedes_benz': 8485,
 'audi': 9093,
 'ford': 3652}

Out of 5 brands:
- Audi, Mercedes_benz, BMW are the most expensive, on average
- Opel & Ford are less expensive 
- Volkswagen is in between the two category, which confirms its popularity

### Exploring Mileage by Brand

In [30]:
brand_mean_mileage = {}

for brand in common_brand:
    mean_mil = autos[autos['brand'] == brand]['odometer_km'].mean()
    brand_mean_mileage[brand] = int(mean_mil)

brand_info = pd.DataFrame(pd.Series(brand_mean_mileage), columns = ['mean_mileage'])
brand_info

mean_mileage
volkswagen           128724
bmw                  132431
opel                 129223
mercedes_benz        130856
audi                 129287
ford                 124068

In [31]:
brand_info['mean_price'] = pd.Series(brand_mean_prices)
brand_info.sort_values(ascending = False, by = 'mean_price')

mean_mileage  mean_price
audi                 129287        9093
mercedes_benz        130856        8485
bmw                  132431        8102
volkswagen           128724        5231
ford                 124068        3652
opel                 129223        2876

Among six most popular car brands listed on German Ebay site, the top 3 are `Audi`, `Mercedes_Benz`, `BMW`. They are, on average, more expensive compare to the rest, but also come with higher mileage, again, on average.

In [32]:
top3 = brand_info.loc[['audi', 'mercedes_benz','bmw']]
top3_mean_mil = top3['mean_mileage'].mean()
print("Average mileage of top 3 most popular German brands: ", top3_mean_mil , " km")

top3_mean_price = top3['mean_price'].mean()
print("Average price of top 3 most popular German brands: ", top3_mean_price , " dollar")

Average mileage of top 3 most popular German brands:  130858.0  km
Average price of top 3 most popular German brands:  8560.0  dollar


**Data cleaning next steps**:
* Identify categorical data that uses german words, translate them and map the values to their english counterparts
* Convert the dates to be uniform numeric data, so `2016-03-21` becomes the integer `20160321`.
* See if there are particular keywords in the name column that you can extract as new columns



**Analysis next steps**:
* Find the most common brand/model combinations
* Split the odometer_km into groups, and use aggregation to see if average prices follows any patterns based on the mileage.
* How much cheaper are cars with damage than their non-damaged counterparts?

#### Covert German words in categorical columns into English

In [33]:
autos_german_to_eng = {}

In [34]:
cate_col = ['seller', 'offer_type', 'vehicle_type','gearbox','fuel_type']
for col in cate_col:
    print(col + ' column:')
    print(list(autos[col].unique()))
    print('\n')

seller column:
['privat', 'gewerblich']


offer_type column:
['Angebot', 'Gesuch']


vehicle_type column:
['bus', 'limousine', 'kleinwagen', 'kombi', nan, 'coupe', 'suv', 'cabrio', 'andere']


gearbox column:
['manuell', 'automatik', nan]


fuel_type column:
['lpg', 'benzin', 'diesel', nan, 'cng', 'hybrid', 'elektro', 'andere']




In [36]:
autos_german_to_eng = {'privat':'private', 'gewerblich':'business', 'Angebot':'offer','Gesuch':'request','manuell': 'manual'
                       ,'automatik':'auto'}

In [37]:
autos['seller'] = autos['seller'].map(autos_german_to_eng)
            
autos['seller'].unique()

array(['private', 'business'], dtype=object)

In [38]:
autos['offer_type'] = autos['offer_type'].map(autos_german_to_eng)

autos['offer_type'].unique()

array(['offer', 'request'], dtype=object)

#### Covert dates to numeric data

In [ ]:
lst = ['date_crawled', 'ad_created', 'last_seen']
autos[lst].head()

In [ ]:
def join_str(date):
    return ''.join(date[:10].split('-'))

In [ ]:
for date in lst:
    autos[date] = autos[date].apply(join_str)
    autos[date] = autos[date].astype('int')

In [ ]:
autos[lst].head()